### DISCUSSION TITLE
Create SVG QR Codes with embedded logos

### DISCUSSION THREAD

**Discussion by nicolomonili** [chillerlan/php-qrcode#244] [Q&A] [answered]
Hi, first of all thanks for the library.

I'm working with this example (https://github.com/chillerlan/php-qrcode/blob/v5.0.x/examples/svgWithLogo.php) and it works fine with all the svg images I download from here https://simpleicons.org/.

But when I try to embed my own SVG logos, or this wikipedia example (https://upload.wikimedia.org/wikipedia/en/8/80/Wikipedia-logo-v2.svg), the QrCode doesn't add my SVG logo (there remains an empty space in the center).

![Screenshot 2024-02-21 alle 22 37 45](https://github.com/chillerlan/php-qrcode/assets/2778813/8f33dfef-7c1d-4b61-85b3-b13d5d067ac2)

Do the SVG images to be embedded have to have a certain format?

Thank you

**Comment by nicolomonili** [comment]
With github logo from https://simpleicons.org/

![Screenshot 2024-02-21 alle 22 39 47](https://github.com/chillerlan/php-qrcode/assets/2778813/976c50b5-54c6-41d0-9385-00d321845712)

**Comment by codemasher** [comment, marked as answer]
Hey, I just tried here on local and ended up with this (ignoring the parse errors that are produced by the wikipedia logo's headers) 

![grafik](https://github.com/chillerlan/php-qrcode/assets/592497/b2540b0e-e720-407b-8d3d-bbf70e260e28)

When you remove the XML and DOCTYPE header lines as well as the fixed width and height from the logo SVG, it works perfectly fine (SVG source below):


![svgWithLogo](https://github.com/chillerlan/php-qrcode/assets/592497/b4a2fdfe-c4ad-4288-8959-5261ecff8ff6)

**Comment by nicolomonili** [comment]
Thanks for the reply. 

I tried to modify the svg code of the wikipedia svg image, keeping only the content of the SVG tag and the following as tag parameters: <svg version="1.1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 103 94" role="img">, but the image is still not displayed. I attach the modified svg image.

![wiki2](https://github.com/chillerlan/php-qrcode/assets/2778813/2b42a3eb-3248-4d7e-ab22-ccbd5191e603)

There is clearly something wrong with the code of the modified svg image.

**Reply by codemasher** [reply]
Try this one:

![Wikipedia logo version 2](https://github.com/chillerlan/php-qrcode/assets/592497/ef01a3ac-94a3-424e-bd71-ad7135ee5ece)

**Reply by nicolomonili** [reply]
It works perfectly with your version.

**Reply by codemasher** [reply]
Keeping the `viewBox` of the logo SVG is crucial - it defines the canvas size of the image.

**Reply by codemasher** [reply]
This is what the final rendered code looks like:

```svg
<g transform="translate(16.875 16.875) scale(0.25)" class="dark">
	<svg version="1.1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 103 94">
	<title>Wikipedia logo version 2</title>

	<!-- logo svg here -->

	</svg>
</g>
```

**Reply by nicolomonili** [reply]
In the meantime, thank you very much, but I haven't yet had the chance to check what you changed compared to the original svg.
I'll try with another svg image (which also didn't work).

**Reply by nicolomonili** [reply]
From the original svg, you have remove this (I can't post this code)

![Screenshot 2024-02-22 alle 16 43 23](https://github.com/chillerlan/php-qrcode/assets/2778813/ed9ab55b-c030-49f4-aa93-1e0335619e3d)

changed the svg node from 

```
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" x="0px" y="0px" width="103px" height="94px" viewBox="0 0 103 94">
```

to

`<svg version="1.1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 103 94">`

and removed all the
`style=""`

it's correct?

**Reply by codemasher** [reply]
Ok, so I saved the file through the browser as .svg - I noticed that for some reason saving it this way removes some of the `<!ENTITY>` tags from the original source. I then ran it through the code reformatter and removed the first 2 lines (xml header and doctype) as they would render the SVG invalid:

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
```

and i removed the `width` and `height` as well as the `x` and `y` attributes from the `<SVG>` element:

`<svg version="1.1" xmlns="http://www.w3.org/2000/svg" x="0px" y="0px" width="103px" height="94px" viewBox="0 0 103 94">`

to

`<svg version="1.1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 103 94">`

this should normally do the trick with any SVG - might need to play areound with custom namespaces maybe - i haven't tested that yet.

**Reply by nicolomonili** [reply]
I found a very useful online tool, which is used to "clean up" the svg image.
After uploading an svg image, you can remove some parameters from the image code (such as the doctype, xml header, style attributes, etc etc..)

https://svgomg.net/

**Comment by nutthakun** [comment]
https://github.com/chillerlan/php-qrcode/blob/main/examples/svgWithLogo.php

I use this code but no logo appears and no errors. 
What do I need to fix?

please .
![image](https://github.com/chillerlan/php-qrcode/assets/25877787/89a2a5a6-962c-4777-8273-dfa1ef79fdad)

**Reply by codemasher** [reply]
You're using the example from `dev-main`, see https://github.com/chillerlan/php-qrcode/discussions/241#discussioncomment-8461662

**Reply by nutthakun** [reply]
Thank you.
